### PR TITLE
[SPARK-32827][SQL] Add spark.sql.maxMetadataStringLength config

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Add a new config `spark.sql.maxMetadataStringLength`. This config aims to limit metadata value length, e.g. file location.

### Why are the changes needed?

Some metadata have been abbreviated by `...` when I tried to add some test in `SQLQueryTestSuite`. We need to replace such value to `notIncludedMsg`. That caused we can't replace that like location value by `className` since the `className` has been abbreviated.

Here is a case:
```
CREATE table  explain_temp1 (key int, val int) USING PARQUET;

EXPLAIN EXTENDED SELECT sum(distinct val) FROM explain_temp1;

-- ignore parsed,analyzed,optimized
-- The output like
== Physical Plan ==
*HashAggregate(keys=[], functions=[sum(distinct cast(val#x as bigint)#xL)], output=[sum(DISTINCT val)#xL])
+- Exchange SinglePartition, true, [id=#x]
   +- *HashAggregate(keys=[], functions=[partial_sum(distinct cast(val#x as bigint)#xL)], output=[sum#xL])
      +- *HashAggregate(keys=[cast(val#x as bigint)#xL], functions=[], output=[cast(val#x as bigint)#xL])
         +- Exchange hashpartitioning(cast(val#x as bigint)#xL, 4), true, [id=#x]
            +- *HashAggregate(keys=[cast(val#x as bigint) AS cast(val#x as bigint)#xL], functions=[], output=[cast(val#x as bigint)#xL])
               +- *ColumnarToRow
                  +- FileScan parquet default.explain_temp1[val#x] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex[file:/home/runner/work/spark/spark/sql/core/spark-warehouse/org.apache.spark.sq...], PartitionFilters: ...
```

### Does this PR introduce _any_ user-facing change?

No, a new config.

### How was this patch tested?

new test.

### CODE REVIEW

**Comment by ulysses-you** [comment]
@maropu @dongjoon-hyun @cloud-fan do you have time to look this ? thanks !

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -2370,6 +2370,13 @@ object SQLConf {
       "(nonnegative and shorter than the maximum size).")
     .createWithDefaultString(s"${ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH}")
 
+  val MAX_METADATA_STRING_LENGTH = buildConf("spark.sql.maxMetadataStringLength")
+    .doc("Maximum number of characters to output for a metadata string. e.g. " +
+      "`DataSourceScanExec`, every value will be abbreviated if exceed length.")
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2375]: `e.g. DataSourceScanExec` => `e.g. file location in DataSourceScanExec`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -2370,6 +2370,13 @@ object SQLConf {
...
+    .version("3.1.0")
+    .intConf
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2377]: plz add `checkValue`

File: sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala [hunk 3]
```diff
@@ -881,6 +882,28 @@ class FileBasedDataSourceSuite extends QueryTest
       }
     }
   }
+
+  test("SPARK-32827: Add spark.sql.maxMetadataStringLength config") {
+    withTempDir { dir =>
+      val tableName = "t1"
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala:888]: nit: `t1` -> `t`

File: sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala [hunk 4]
```diff
@@ -881,6 +882,28 @@ class FileBasedDataSourceSuite extends QueryTest
...
+      val path = s"${dir.getCanonicalPath}/$tableName"
+      withTable(tableName) {
+        sql(s"create table t1(c int) using parquet location '$path'")
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala:891]: plz use uppercases for SQL keywords, e.g., `CREATE TABLE`

**Comment by maropu** [comment]
Adding the config looks okay to me.

File: sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala (hunk 4 above)
Comment (ulysses-you) [reply sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala]: done.

**Comment by xkrogen** [comment]
Glad to see this happening @ulysses-you ! We have a similar internal patch we added recently that we've been planning to open source. Nice to see that the changes are even simpler on the 3.x line.

File: sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala [hunk 5]
```diff
@@ -881,6 +882,28 @@ class FileBasedDataSourceSuite extends QueryTest
       }
     }
   }
+
+  test("SPARK-32827: Add spark.sql.maxMetadataStringLength config") {
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala:886]: Nit: `Add spark.sql.maxMetadataStringLength config` => `Set max metadata string length`

**Comment by gengliangwang** [comment]
@ulysses-you  thanks for the work.
It seems that the PR only changes the file source related metadata. Is there other places we can use the new config as well?

**Comment by ulysses-you** [comment]
Thanks @gengliangwang .The origin propose is for explain test what I said in pr description, related pr is [#29586](https://github.com/apache/spark/pull/29586). Location value has been abbreviated by `...` since default metadata length is 100. We can add the config at head of test file like `--SET spark.sql.maxMetadataStringLength=500;` if we have this config, then we can get the total location value.

I'm not sure if it's well to do these change in one pr. But I can merge them into one pr if you think it's ok.

**Comment by AngersZhuuuu** [comment]
> @ulysses-you thanks for the work.
> It seems that the PR only changes the file source related metadata. Is there other places we can use the new config as well?

I will use this config in my pr https://github.com/apache/spark/pull/29739.
Similar work in my pr and reverted waiting this.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by ulysses-you** [comment]
Thanks for merging, thanks all !

**Pull request closed without merging**